### PR TITLE
Fixes remaining issues with DynamicDependency attribute and adds tests

### PR DESCRIPTION
**Pull request by vitek-karas**
Ports over all DynamicDependencyAttribute tests we have in linker. All tests are now passing, two had to be partially disabled due to:
* copyused testing which NativeAOT doesn't really support
* known missing support for XML attribute files in AOT

Product changes:
* DynamicDependencyAttribute will produce a warning if it can't resolve the target of the dependency.
* Support for DynamicDependencyAttribute on fields

Small improvements to assembly checker - mainly ability to handle delegate types.

### CODE REVIEW

**Review by sbomer — APPROVED** [review]
LGTM, thank you!

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/DynamicDependencyAttributeAlgorithm.cs [hunk 1]
```diff
@@ -26,135 +30,187 @@ public static void AddDependenciesDueToDynamicDependencyAttribute(ref Dependency
...
-                            continue;
-                        }
-
-                        targetType = DocumentationSignatureParser.GetTypeByDocumentationSignature((IAssemblyDesc)asm, typeStringFromAttribute);
-                        if (targetType == null)
-                        {
-                            // _context.LogWarning ($"Unresolved type '{typeName}' in DynamicDependencyAttribute", 2036, context);
-                            continue;
-                        }
+                        (factory.MetadataManager as UsageBasedMetadataManager)?.Logger?.LogWarning(
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/DynamicDependencyAttributeAlgorithm.cs:91]: Is the `is` needed? Basically if we run this with other than the UsageBased manager, we're running this logic twice (first time scanner, second time real compilation). We should ideally not do that - it's useless work the second time.

```suggestion
                        ((UsageBasedMetadataManager)factory.MetadataManager).Logger.LogWarning(
```

**Review by MichalStrehovsky — COMMENTED** [review]
I think there is one more issue remaining - will this keep XXX for the below? I think we need to also call into this algorithm from `FieldMetadataNode`.

```
typeof(Foo).GetFields();

class Foo
{
    [DynamicDependency(XXX)]
    public int Field;
}
```

**Comment by MichalStrehovsky** [comment]
> I think there is one more issue remaining - will this keep XXX for the below? I think we need to also call into this algorithm from `FieldMetadataNode`.
> 
> ```
> typeof(Foo).GetFields();
> 
> class Foo
> {
>     [DynamicDependency(XXX)]
>     public int Field;
> }
> ```

And thinking about it more, also from `MethodMetadataNode` to cover:

```csharp
typeof(Foo).GetMethods();

class Foo
{
    [DynamicDependency(XXX)]
    public abstract void M();
}
```

File: src/coreclr/tools/aot/Mono.Linker.Tests.Cases/DynamicDependencies/Dependencies/AssemblyDependency.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Mono.Linker.Tests.Cases.DynamicDependencies.Dependencies
```
Comment (marek-safar) [inline src/coreclr/tools/aot/Mono.Linker.Tests.Cases/DynamicDependencies/Dependencies/AssemblyDependency.cs:1]: Do we want to duplicate all these tests? Wouldn't be better to have that in some shared location?

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/DynamicDependencyAttributeAlgorithm.cs [hunk 3]
```diff
@@ -26,135 +30,187 @@ public static void AddDependenciesDueToDynamicDependencyAttribute(ref Dependency
...
+            // * A string that contains a documentation signature
+            // * DynamicallyAccessedMembers enum
 
-                if (fixedArgs.Length > 0 && fixedArgs[0].Value is string sigFromAttribute)
+            var fixedArgs = attribute.FixedArguments;
+            TypeDesc targetType;
+
+            if (fixedArgs.Length > 0 && fixedArgs[0].Value is string sigFromAttribute)
+            {
+                if (fixedArgs.Length == 1)
```
Comment (marek-safar) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/DynamicDependencyAttributeAlgorithm.cs:74]: nit: use switch

**Comment by vitek-karas** [comment]
> And thinking about it more, also from MethodMetadataNode

Technically that should not be needed since reflection enabled method is automatically callable and thus will have an entry point - which is what the current logic is tied to. (I added a simple test case similar to your example and it already works).
I'll look into this a little bit more though.

**Comment by MichalStrehovsky** [comment]
> Technically that should not be needed since reflection enabled method is automatically callable and thus will have an entry point

Abstract methods don't have an entrypoint, but it's possible we're already taking the right codepaths. Maybe a test would still be useful if we don't already have such test.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/DynamicDependencyAttributeAlgorithm.cs (hunk 1 above)
Comment (vitek-karas) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/DynamicDependencyAttributeAlgorithm.cs:91]: Changed

File: src/coreclr/tools/aot/Mono.Linker.Tests.Cases/DynamicDependencies/Dependencies/AssemblyDependency.cs (hunk 2 above)
Comment (marek-safar) [reply src/coreclr/tools/aot/Mono.Linker.Tests.Cases/DynamicDependencies/Dependencies/AssemblyDependency.cs:1]: I think if you put the files in the right location it'd be enough and save us from deleting them all in few weeks/months

Comment (vitek-karas) [reply src/coreclr/tools/aot/Mono.Linker.Tests.Cases/DynamicDependencies/Dependencies/AssemblyDependency.cs:1]: I think the tricky bit is having them in the right location (don't think we agreed on one yet), and I don't want to have that discussion as part of a random unrelated PR.

**Comment by vitek-karas** [comment]
All the failures are known.

**Pull request merged**